// @histoire/app/dist/bundled/components/misc/CheckerboardPattern.vue.js
import { createElementBlock, openBlock, createElementVNode } from "@histoire/vendors/vue";
import _export_sfc from "../../_virtual/_plugin-vue_export-helper.js";
const _sfc_main = {};
const _hoisted_1 = { fill: "none" };
function _sfc_render(_ctx, _cache) {
  return openBlock(), createElementBlock("svg", _hoisted_1, _cache[0] || (_cache[0] = [
    createElementVNode("pattern", {
      id: "checkerboard",
      width: "64",
      height: "64",
      patternUnits: "userSpaceOnUse"
    }, [
      createElementVNode("rect", {
        x: "0",
        y: "0",
        width: "32",
        height: "32",
        fill: "currentColor"
      }),
      createElementVNode("rect", {
        x: "32",
        y: "32",
        width: "32",
        height: "32",
        fill: "currentColor"
      })
    ], -1),
    createElementVNode("rect", {
      x: "0",
      y: "0",
      width: "100%",
      height: "100%",
      fill: "url(#checkerboard)"
    }, null, -1)
  ]));
}
const CheckerboardPattern = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export {
  CheckerboardPattern as default
};
